apps/emqx_schema_registry/src/emqx_schema_registry_config.erl

```diff
@@ -314,3 +316,152 @@ convert_certs(Name, #{<<"parameters">> := #{<<"ssl">> := SSL} = Params0} = RawCo
     end;
 convert_certs(_Name, RawConf) ->
     {ok, RawConf}.
+
+convert_protobuf_bundle(
+    Name,
+    #{
+        <<"type">> := <<"protobuf">>,
+        <<"source">> := #{
+            <<"type">> := <<"bundle">>,
+            <<"files">> := _
+        }
+    } = RawConf0
+) ->
+    do_convert_protobuf_bundle(Name, RawConf0);
+convert_protobuf_bundle(_Name, RawConf) ->
+    {ok, RawConf}.
+
+do_convert_protobuf_bundle(Name, RawConf0) ->
+    #{
+        <<"type">> := <<"protobuf">>,
+        <<"source">> := #{
+            <<"type">> := <<"bundle">>,
+            <<"files">> := Files
+        }
+    } = RawConf0,
+    maybe
+        {ok, RootFile, OtherFiles} ?= find_protobuf_bundle_root_file(Files),
+        ok ?= validate_protobuf_path(RootFile),
+        {ok, RootPath} ?= copy_protobuf_bundle_to_data_dir(Name, RootFile, OtherFiles),
+        ok ?= validate_protobuf_imports({path, RootPath}),
+        NewSource = #{
+            <<"type">> => <<"bundle">>,
+            <<"root_proto_path">> => RootPath
+        },
+        RawConf = RawConf0#{<<"source">> := NewSource},
+        {ok, RawConf}
+    else
+        {error, {invalid_or_missing_imports, _} = Reason} ->
+            delete_protobuf_bundle_data_dir(Name),
```

Comment:
should be in a try after clause ?
i.e. purge this dir if failed.

Reply:
should only purge if failed, not always, hence this clause; `after` would always purge

Reply:
should only delete new files...  if updating an existing schema, should not delete existing files if new schema fails to import.

Reply:
sorry, you are right, we should not delete in the 'after' clause.

I guess what I wanted to say is purge in case of failure

> should only delete new files

to make it simple, i guess we can either

- do not allow update, force delete first.
- use a temporary dir for the new schemas, delete the old and then rename (also lowers the chance of half-updated directory in case crashed in the middle)

Reply:
> use a temporary dir for the new schemas, delete the old and then rename (also lowers the chance of half-updated directory in case crashed in the middle)

I was thinking of this approach;  there's still a small chance of being left in a broken state, which I was trying to avoid, but I guess we can't.  :crying_cat_face: 

> to make it simple: do not allow update, force delete first.

I think we can't dismiss this case here: when importing config backups, we need to overwrite what's there.  :thinking: